.mocharc.js

```diff
@@ -11,7 +11,8 @@ module.exports = {
   recursive: true,
   timeout: (process.env.CIRCLECI === 'true' ? 5 : 2) * 1000, // Circle CI has low-performance CPUs.
   reporter: 'dot',
-  require: [require.resolve('./test/utils/setupBabel'), require.resolve('./test/utils/setupJSDOM')],
+  'node-option': ['loader=tsx/cjs'],
```

Comment:
Would it continue to run babel.config.js? IMHO we need the test to run the same Babel plugins as run when publishing to npm.

Reply:
> Would it continue to run babel.config.js? IMHO we need the test to run the same Babel plugins that run when publishing to npm.

It wouldn't, personally I don't really see how much value it adds. We run these mocha tests in a completely different environment than our users anyway. i.e. native Node.js, while most of our users run it in browsers through a bundler, that compiles each file down to their target env anyway.

It already doesn't run the same plugins by the way, looking at the amount of branching that is happening. + it runs tests for only one babel configuration while we publish multiple different builds. If anything, for the long term I'd advocate for progressively moving our builds away from babel. Too slow and too much complexity in 2024 🙂.

Reply:
>  I don't really see how much value it adds.

E.g. that https://github.com/merceyz/babel-plugin-optimize-clsx doesn't introduce bugs

Reply:
> I want transpilation capabilities, one way or another.

Ok, I'll look for another way to make this work with babel then.

Reply:
@oliviertassinari `@babel/register` [can't handle pure ESM modules](https://babeljs.io/docs/babel-register#experimental-babel-8-implementation) like d3-* packages it seems. But we can just run both babel and tsx I guess, the latter seems to be able to handle it just fine, not sure what babel is talking about.

This installs both the babel and tsx hooks, which is somewhat redundant, but it's the best I can offer if we want to run babel plugins.

Reply:
Getting rid of Babel can let us use SWC for Next JS build, potentially speeding it up substantially (which, then, can turn into lower CI operating costs), so it seems like it's a direction we should at least consider.

Reply:
Also noticed that babel is being dropped in favour for `tsup` in zero-config runtime https://github.com/mui/material-ui/pull/40426#issuecomment-1876938960

Reply:
>  it's like asking for a banana and getting the gorilla holding it, together with the whole jungle

Haha.

I think that the problem is that our users don't care if our dev tools are slow as long as their components are performant in production. We rely on a couple of Babel plugins to make this work. https://github.com/oliviertassinari/babel-plugin-transform-react-remove-prop-types probably has the most impact in terms of improving DX without harming bundle size. Before TypeScript was a thing, that plugin used to be popular (as a % of react-dom):

<img src="https://github.com/alexfauquette/mui-x/assets/3165635/995e0ee0-66d3-4bd7-bc7f-90c0e7c4b0b0" width="651">

https://npm-stat.com/charts.html?package=react-dom&package=babel-plugin-transform-react-remove-prop-types&from=2020-01-08&to=2024-01-08

But in any cases, the zero runtime is using Babel to do the transpilation, so sounds like we are digging our hole deeper 😄

Reply:
> I think that the problem is that our users don't care if our dev tools are slow as long as their components are performant in production. We rely on a couple of Babel plugins to make this work.
> https://github.com/oliviertassinari/babel-plugin-transform-react-remove-prop-types probably has the most impact in terms of improving DX without harming bundle size. Before TypeScript was a thing, that plugin used to be popular (as a % of react-dom):

It is not only about slowness. It is also about being able to use certain features. e.g. Next.js+babel doesn't support typescript `satisfies`. I've noticed instances in the X codebase where `satisfies` would lead to better typings. And end-users do care about that. e.g. As a way to better narrow down column definitions for certain column types. The problem is not that we're using babel, it's that we've locked ourselves in it. With very little benefit, things that can be done with techniques that exist in every compiler. e.g. variable replacement and dead code elimination:

```tsx
const Baz = (props) => <div {...props} />;

if (process.env.NODE_ENV === 'development') {
  Baz.propTypes = {
    className: PropTypes.string,
  };
}
```

I'm totally fine to use babel as a compiler where appropriate, I just wish we were able to use other tools as well. Now the slowness of babel dictates the speed of every tool we use.

> But in any cases, the zero runtime is using Babel to do the transpilation, so sounds like we are digging our hole deeper 😄.

It's built on top of babel, but it's compiled using esbuild. They used the most appropriate tool for the job. babel because easily extendible with javascript. esbuild because it's fast.

> It seems that only Next.js is pushing the SWC narrative: https://npm-stat.com/charts.html?package=%40swc%2Fcore&package=%40babel%2Fcore&package=next&from=2023-01-08&to=2024-01-08.

Yep, vite is pushing the esbuild narrative, look where the growth is https://npm-stat.com/charts.html?package=%40swc%2Fcore&package=%40babel%2Fcore&package=esbuild&from=2023-01-08&to=2024-01-08
The point is not to move to SWC across the repo, it's about being able to use SWC in Next.js and babel or tsup to compile our bundles. It's about keeping the source code as close to the standards as possible to have it as standard as possible.

> And wow, it looks like Vite is pushing the Rollup narrative, I wouldn't have expected Rollup to take over Webpack https://npm-stat.com/charts.html?package=webpack&package=rollup&package=vite&from=2023-01-08&to=2024-01-08. We will see if Turbopack changes this.

vite is porting rollup to rust as we speak (https://vitejs.dev/blog/announcing-vite5) and planning to move their whole toolchain to Rust. Vite + React is primed to win 2024 https://npm-stat.com/charts.html?package=vite&package=%40vitejs%2Fplugin-react&package=next&from=2022-01-07&to=2024-01-07. We should prioritize https://github.com/vitejs/vite/issues/12423 if we want to stay relevant with those users.

Reply:
Ok, this all makes sense to me.

Reply:
Whoops, rollup is pushing the swc narrative after all, it just doesn't reflect in the downloads as they seem to compile from source: https://github.com/rollup/rollup/pull/5073